### PR TITLE
Support TautomerQuery and MolBundle queries in the cartridge

### PR DESCRIPTION
**Pull request by greglandrum**
The major work here is to add support for using `TautomerQuery`s and `MolBundle`s (like what you get from the MolEnumerator) as queries in the cartridge. This uses the functionality added in #6443 (and fixes a few bugs in that). 

Overview of what's here:
- adds an xqmol (extended query molecule) type to the cartridge which can be used in substruct queries with molecules.
- add `mol_to_xqmol()` to the cartridge

Other known shortcomings, I don't think all of these need to be resolved for a merge:
- No integration with the indexing system. Making this work has eluded me so far.

### CODE REVIEW

**Comment by greglandrum** [comment]
@bjonnh-work : you may be interested in this work too.

File: Code/PgSQL/rdkit/mol_op.c [hunk 1]
```diff
@@ -407,6 +440,69 @@ Datum mol_to_svg(PG_FUNCTION_ARGS) {
...
+}
+
+PGDLLEXPORT Datum xqmol_to_tautomerquery(PG_FUNCTION_ARGS);
+PG_FUNCTION_INFO_V1(xqmol_to_tautomerquery);
+Datum xqmol_to_tautomerquery(PG_FUNCTION_ARGS) {
+  CXQMol xqmol;
+  fcinfo->flinfo->fn_extra =
+      searchXQMolCache(fcinfo->flinfo->fn_extra, fcinfo->flinfo->fn_mcxt,
+                     PG_GETARG_DATUM(0), NULL, &xqmol, NULL);
+  Assert(mol != 0);
```
Comment (rvianello) [inline Code/PgSQL/rdkit/mol_op.c:471]: I think line 471 should probably be `Assert(xqmol != 0);`

File: Code/PgSQL/rdkit/adapter.cpp [hunk 2]
```diff
@@ -2277,3 +2282,166 @@ extern "C" char *findMCS(void *vmols, char *params) {
...
+}
+
+extern "C" void freeCXQMol(CXQMol data) {
+  auto *mol = (ExtendedQueryMol *)data;
+  delete mol;
+}
+
+extern "C" CXQMol MolToXQMol(CROMol m, bool doEnumeration, bool doTautomers,
+                             bool adjustQueryProperties, const char *params) {
+  const ROMol *im = (ROMol *)m;
```
Comment (ptosco) [inline Code/PgSQL/rdkit/adapter.cpp:2393]:
```suggestion
  auto *im = (const ROMol *)m;
```

File: Code/PgSQL/rdkit/adapter.cpp [hunk 3]
```diff
@@ -2277,3 +2282,166 @@ extern "C" char *findMCS(void *vmols, char *params) {
   // elog(WARNING, "findMCS(): molecules deleted. FINISHED.");
   return strdup(mcs.c_str());
 }
+
+extern "C" char *makeXQMolBlob(CXQMol data, int *len) {
+  StringData.clear();
```
Comment (ptosco) [inline Code/PgSQL/rdkit/adapter.cpp:2287]:
```suggestion
  PRECONDITION(len, “”);
  StringData.clear();
```

**Review by ptosco — APPROVED** [review]
LGTM. I only added a few suggestions, none of which is binding.

**Pull request merged**